### PR TITLE
Migrating rustc_infer to session diagnostics (part 5)

### PR DESCRIPTION
**Pull request by IntQuant**
@rustbot label +A-translation
cc https://github.com/rust-lang/rust/issues/100717

### CODE REVIEW

**Comment by IntQuant** [comment]
r? rust-lang/diagnostics

File: compiler/rustc_infer/src/errors/mod.rs [hunk 1]
```diff
@@ -1157,3 +1145,388 @@ pub struct OpaqueCapturesLifetime<'tcx> {
     pub opaque_ty_span: Span,
     pub opaque_ty: Ty<'tcx>,
 }
+
+pub struct DiagArg<T>(pub T);
```
Comment (davidtwco) [inline compiler/rustc_infer/src/errors/mod.rs:1149]: Why is this type necessary? Could [`rustc_errors::diagnostic_impls::DiagnosticArgFromDisplay`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/diagnostic_impls/struct.DiagnosticArgFromDisplay.html#) be used instead?

File: compiler/rustc_infer/src/errors/mod.rs [hunk 2]
```diff
@@ -1157,3 +1145,388 @@ pub struct OpaqueCapturesLifetime<'tcx> {
...
+        applicability = "maybe-incorrect"
+    )]
+    CastBoth {
+        #[primary_span]
+        span: Span,
+        #[skip_arg]
+        fn_name: String,
+        #[skip_arg]
+        found_sig: Binder<'a, FnSig<'a>>,
+        expected_sig: DiagArg<Binder<'a, FnSig<'a>>>,
```
Comment (davidtwco) [inline compiler/rustc_infer/src/errors/mod.rs:1224]: Presumably `DiagArg` is necessary for `Binder`? Can `IntoDiagnosticArg` not be implemented for this type?

File: compiler/rustc_infer/src/errors/mod.rs [hunk 3]
```diff
@@ -1157,3 +1145,388 @@ pub struct OpaqueCapturesLifetime<'tcx> {
...
+}
+
+pub struct SuggestTuplePatternMany {
+    pub path: String,
+    pub cause_span: Span,
+    pub compatible_variants: Vec<String>,
+}
+
+impl AddToDiagnostic for SuggestTuplePatternMany {
+    fn add_to_diagnostic_with<F>(self, diag: &mut rustc_errors::Diagnostic, f: F)
```
Comment (davidtwco) [inline compiler/rustc_infer/src/errors/mod.rs:1344]: Can this be a `Vec` of a type that has `#[multipart_suggestion(..)]`?

File: compiler/rustc_infer/src/errors/mod.rs [hunk 4]
```diff
@@ -1157,3 +1145,388 @@ pub struct OpaqueCapturesLifetime<'tcx> {
...
+                    (self.cause_span.shrink_to_hi(), ")".to_string()),
+                ]
+            }),
+            rustc_errors::Applicability::MaybeIncorrect,
+        );
+    }
+}
+
+#[derive(Subdiagnostic)]
+pub enum Error0308Subdiags {
```
Comment (davidtwco) [inline compiler/rustc_infer/src/errors/mod.rs:1364]: Could this have a more descriptive name?

File: compiler/rustc_infer/src/errors/mod.rs [hunk 5]
```diff
@@ -1157,3 +1145,388 @@ pub struct OpaqueCapturesLifetime<'tcx> {
...
+        code = "let "
+    )]
+    AddLetForLetChains {
+        #[primary_span]
+        span: Span,
+    },
+}
+
+#[derive(Diagnostic)]
+pub enum FailureCodeDiagnostics {
```
Comment (davidtwco) [inline compiler/rustc_infer/src/errors/mod.rs:1432]: Could this have a more descriptive name too?

**Review by davidtwco — CHANGES_REQUESTED** [review]

File: compiler/rustc_infer/src/errors/mod.rs (hunk 3 above)
Comment (IntQuant) [reply compiler/rustc_infer/src/errors/mod.rs:1344]: Macro doesn't support a dynamic list of different suggestions, even tho all of them are on the same span([zulip link](https://rust-lang.zulipchat.com/#narrow/stream/336883-i18n/topic/.23100717.20deriving.20multipart.20suggestions)), so no.

File: compiler/rustc_infer/src/infer/error_reporting/mod.rs [hunk 6]
```diff
@@ -2828,57 +2791,89 @@ impl<'tcx> InferCtxt<'tcx> {
 }
 
 pub enum FailureCode {
```
[the remark this answers is not in the document]
Comment (IntQuant) [reply compiler/rustc_infer/src/infer/error_reporting/mod.rs:2793]: It's actually still used in [rustc_hir_typeck](https://github.com/IntQuant/rust/blob/1905f1972647120da01a48b23f9eba0a30ea2f56/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs#L767)

File: compiler/rustc_infer/src/infer/error_reporting/mod.rs [hunk 7]
```diff
@@ -1888,232 +1887,196 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
...
-        trace: TypeTrace<'tcx>,
+        trace: &TypeTrace<'tcx>,
         terr: TypeError<'tcx>,
-    ) -> DiagnosticBuilder<'tcx, ErrorGuaranteed> {
+    ) -> Vec<Error0308Subdiags> {
         use crate::traits::ObligationCauseCode::MatchExpressionArm;
-
-        debug!("report_and_explain_type_error(trace={:?}, terr={:?})", trace, terr);
-
+        fn escape_literal(s: &str) -> String {
```
[the remark this answers is not in the document]
Comment (IntQuant) [reply compiler/rustc_infer/src/infer/error_reporting/mod.rs]: Sure!

File: compiler/rustc_infer/src/errors/mod.rs (hunk 5 above)
Comment (IntQuant) [reply compiler/rustc_infer/src/errors/mod.rs:1432]: Done

File: compiler/rustc_infer/src/errors/mod.rs (hunk 4 above)
Comment (IntQuant) [reply compiler/rustc_infer/src/errors/mod.rs:1364]: Fixed

Comment (IntQuant) [reply compiler/rustc_infer/src/errors/mod.rs:1364]: Fixed

File: compiler/rustc_infer/src/errors/mod.rs (hunk 2 above)
Comment (IntQuant) [reply compiler/rustc_infer/src/errors/mod.rs:1224]: Looks like it already was implemented for a specific kind of `Binder` (`Binder<'tcx, ExistentialTraitRef<'tcx>>`), so I just extended it to any `Binder` that is `Display`.

File: compiler/rustc_infer/src/errors/mod.rs (hunk 1 above)
Comment (IntQuant) [reply compiler/rustc_infer/src/errors/mod.rs:1149]: Resolved by implementing `IntoDiagnosticArg` for `Binder`

File: compiler/rustc_infer/src/infer/error_reporting/mod.rs [hunk 8]
```diff
@@ -1888,232 +1887,196 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
...
-                                }
+                    }
+                }
+                // For code `if Some(..) = expr `, the type mismatch may be expected `bool` but found `()`,
+                // we try to suggest to add the missing `let` for `if let Some(..) = expr`
+                (ty::Bool, ty::Tuple(list)) => if list.len() == 0 {
+                    suggestions.extend(self.suggest_let_for_letchains(&trace.cause, span));
+                }
+                (ty::Array(_, _), ty::Array(_, _)) => 'block: {
+                    let hir = self.tcx.hir();
```
[the remark this answers is not in the document]
Comment (IntQuant) [reply compiler/rustc_infer/src/infer/error_reporting/mod.rs]: I've extracted the `(ty::Array(_, _), ty::Array(_, _))` arm, which seems to be the longest.

File: compiler/rustc_middle/src/ty/sty.rs [hunk 9]
```diff
@@ -1159,6 +1148,15 @@ impl<'tcx, T: IntoIterator> Binder<'tcx, T> {
     }
 }
 
+impl<'tcx, T> IntoDiagnosticArg for Binder<'tcx, T>
+where
+    Binder<'tcx, T>: Display,
+{
+    fn into_diagnostic_arg(self) -> DiagnosticArgValue<'static> {
+        self.to_string().into_diagnostic_arg()
```
Comment (davidtwco) [inline compiler/rustc_middle/src/ty/sty.rs:1156]: Should this be `T: IntoDiagnosticArg` and then `self.0.into_diagnostic_arg()`?

**Review by davidtwco — APPROVED** [review]
Apologies for the delay in getting back to this, one last comment then r=me.

**Review by davidtwco — APPROVED** [review]

**Pull request merged**